syntax = "proto3";

package posts;

import "google/protobuf/timestamp.proto";
import "google/protobuf/empty.proto";

message Post {
  int32 id = 1;
  string title = 2;
  string description = 3;
  string creator_id = 4;
  google.protobuf.Timestamp created_at = 5;
  google.protobuf.Timestamp updated_at = 6;
  bool private = 7;
  repeated string tags = 8;
}

message PostRequest {
  int32 id = 1;
  string user_id = 2;
}

message CreatePostRequest {
  string title = 1;
  string description = 2;
  string user_id = 3;
  bool private = 4;
  repeated string tags = 5;
}

message UpdatePostRequest {
  int32 id = 1;
  string title = 2;
  string description = 3;
  string user_id = 4;
  bool private = 5;
  repeated string tags = 6;
}

message ListPostsRequest {
  int32 page = 1;
  int32 limit = 2;
  string user_id = 3;
}

message ListPostsResponse {
  repeated Post post = 1;
  int32 total = 2;
}

service PostService {
  rpc CreatePost(CreatePostRequest) returns (Post);
  rpc GetPost(PostRequest) returns (Post);
  rpc UpdatePost(UpdatePostRequest) returns (Post);
  rpc DeletePost(PostRequest) returns (google.protobuf.Empty);
  rpc ListPosts(ListPostsRequest) returns (ListPostsResponse);
}